crates/iceberg/src/scan.rs

```diff
@@ -151,42 +162,118 @@ impl TableScan {
                         ));
                     }
                     DataContentType::Data => {
-                        file_scan_tasks.push(Ok(FileScanTask {
-                            data_file: manifest_entry.clone(),
+                        plan_files.push(PlanFile {
+                            data_files: manifest_entry.data_file().clone(),
                             start: 0,
-                            length: manifest_entry.file_size_in_bytes(),
-                        }));
+                            len: manifest_entry.file_size_in_bytes(),
+                        });
                     }
                 }
             }
         }
 
-        Ok(iter(file_scan_tasks).boxed())
+        Ok(plan_files)
     }
 }
 
+/// PlanFile is a file to be scanned. It's the smallest unit of scan.
+#[derive(Debug)]
+pub struct PlanFile {
+    data_files: DataFile,
+    start: u64,
+    len: u64,
+}
+
 /// A task to scan part of file.
 #[derive(Debug)]
 #[allow(dead_code)]
 pub struct FileScanTask {
-    data_file: ManifestEntryRef,
-    start: u64,
-    length: u64,
+    plan_files: Vec<PlanFile>,
+    file_io: FileIO,
 }
 
 /// A stream of arrow record batches.
 pub type ArrowRecordBatchStream = BoxStream<'static, crate::Result<RecordBatch>>;
 
 impl FileScanTask {
+    /// Create a new file scan task from `PlanFiles`.
+    pub fn new(plan_files: Vec<PlanFile>, file_io: FileIO) -> Self {
+        Self {
+            plan_files,
+            file_io,
+        }
+    }
+
+    async fn build_stream(
+        plan_file: &PlanFile,
+        file_io: &FileIO,
+    ) -> crate::Result<ArrowRecordBatchStream> {
+        match plan_file.data_files.file_format {
+            crate::spec::DataFileFormat::Parquet => {
+                let parquet_reader = ParquetFileReader {
+                    reader: file_io
+                        .new_input(&plan_file.data_files.file_path)
+                        .unwrap()
+                        .reader()
+                        .await
+                        .unwrap(),
+                };
+                let stream = ParquetRecordBatchStreamBuilder::new(parquet_reader)
```

Comment:
TODO: I'm working on https://github.com/upondal/opendal-parquet-io for best performance.